opta/meister.py

```diff
@@ -0,0 +1,172 @@
+                        "keyword_args": kwds,
+                        "start_time": start_time,
+                        "end_time": end_time,
+                        "id": uuid4(),
+                        "caller_file": "/".join(frame_info.filename.split("/")[-2:]),
+                        "caller_line": frame_info.lineno,
+                    }
+                )
+            print(
+                f"{fg('green')}Meister{attr(0)}: Function {func.__name__} took {end_time - start_time} seconds"
+            )
+            return output
+
+    else:
+
+        @wraps(func)
+        def wrapper(*args: Any, **kwds: Any) -> Any:
+            frame_info = inspect.stack()[1]
+            start_time = time.time()
+            output = func(*args, **kwds)
+            end_time = time.time()
+            if end_time - start_time <= MEISTER_TIME_THRESHOLD:
+                return output
+            MEISTER_RECORDS.append(
+                {
+                    "id": uuid4(),
+                    "function_name": func.__name__,
+                    "positional_args": args,
+                    "keyword_args": kwds,
+                    "start_time": start_time,
+                    "end_time": end_time,
+                    "caller_file": "/".join(frame_info.filename.split("/")[-2:]),
+                    "caller_line": frame_info.lineno,
+                }
+            )
+            print(
+                f"{fg('green')}Meister{attr(0)}: Function {func.__name__} took {round_sig(end_time - start_time)} seconds"
+            )
+            return output
+
+    return wrapper
+
+
+def make_report() -> None:
+    end_time = time.time()
+    total_time = end_time - MEISTER_START_TIME
+    sorted_records = sorted(
+        MEISTER_RECORDS, key=lambda x: x["end_time"] - x["start_time"], reverse=True
+    )
+    sorted_records = sorted_records[:MEISTER_REPORT_SIZE]
+    print(
+        f"{fg('green')}Meister Report{attr(0)}: Total time: {round_sig(total_time)} seconds. Here are the top "
+        f"{min(MEISTER_REPORT_SIZE, len(sorted_records))} longest function calls"
+    )
+    color_map: Dict[UUID, int] = {}
+    for idx, record in enumerate(sorted_records):
+        color_idx = 2 + (idx % 254)
+        positional_args_str = str(record["positional_args"])
+        if len(positional_args_str) > 40:
+            positional_args_str = positional_args_str[:36] + "...)"
+        keyword_args_str = str(record["keyword_args"])
+        if len(keyword_args_str) > 40:
+            keyword_args_str = keyword_args_str[:36] + "...)"
+        print(
+            f"{fg(color_idx)}{record['function_name']}{attr(0)} called in file {record['caller_file']}, line {record['caller_line']} with args {positional_args_str}, "
+            f"kwargs {keyword_args_str}: {round_sig(record['end_time'] - record['start_time'])} seconds"
+        )
+        color_map[record["id"]] = color_idx
+    max_column = os.get_terminal_size()[0]
```

Comment:
```suggestion
    try:
        max_column = os.get_terminal_size()[0]
    except OSError:
        max_column = 80  # Assume 80 columns if we aren't in a terminal
```

`os.get_terminal_size` raises if stdout isn't attached to a terminal.